dagger_fundamentals/build.gradle

```diff
@@ -25,3 +27,22 @@ idea {
     }
 }
 
+static def proguardPath(File jar) {
+    return new File(jar.parentFile, jar.name.replace(".jar", "-pro.jar"))
+}
+// gradlew :dagger_fundamentals:cleanProguard :dagger_fundamentals:proguard --console=plain --info
+// See dagger_fundamentals/build/libs/dagger_fundamentals-pro.jar
+// in Project view (expand jar, open class files and "Decompile to Java")
+def proguard = tasks.register("proguard", ProGuardTask.class) {
+    dependsOn(tasks.jar)
+    configuration project.file("proguard.pro")
+    injars project.jar.archivePath
+    outjars proguardPath(project.jar.archivePath)
+
+    libraryjars "${System.getProperty('java.home')}/lib/rt.jar"
+    libraryjars configurations.runtimeClasspath
+}
+tasks.classes.finalizedBy(proguard)
```

Comment:
Spoke too soon. So it looks like this will mean proguard is run each compilation so, you get the following error each time you build or run of the parts 
 
`Input jars and output jars must be different [/Users/gregm/dev/TrainlingLine/dagger_fundamentals/build/libs/dagger_fundamentals-pro.jar].`

Reply:
That's weird, comment the line out, I'll try to look at the error. The proguard task should still be runnable standalone. I tried many times though in different combos, not sure what's going on yet.

Reply:
Based on my other discovery, proguard will work fine if the dependency below is commented out.